R/data.table.R

```diff
@@ -1942,9 +1942,42 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
       X[[j]] = xj
     }
   }
-  X = unlist(X, recursive = FALSE, use.names = FALSE)
-  dim(X) <- c(n, length(X)/n)
-  dimnames(X) <- list(rownames.value, unlist(collabs, use.names = FALSE))
+  
+  # If any columns of x are multi-column, which can only occur if x is
+  # constructed incorrectly (see test 2074.07), use the old non-C
+  # method of converting to a matrix.
+  col.classes = sapply(X, class)
```

Comment:
`sapply(X, class)` will return a list, i think precisely for the cases you're targeting here, e.g.

```
x = list(data.table(a=1), Sys.time(), b=3L)
sapply(x, class)
# [[1]]
# [1] "data.table" "data.frame"

# [[2]]
# [1] "POSIXct" "POSIXt" 

# $b
# [1] "integer"
```

Reply:
Maybe `sapply(X, NCOL)` would be better?

Reply:
Thanks! I've fixed in the latest commit. I need to know the class of the columns anyway for type conversion, and using unlist(lapply()) will handle other cases we haven't though of where a column may inherit multiple types.